ext/src/ruby_api/memory/slice.rs

```diff
@@ -0,0 +1,193 @@
+use magnus::{
+    memoize, method,
+    r_typed_data::DataTypeBuilder,
+    rb_sys::{AsRawId, AsRawValue, FromRawValue},
+    value::Id,
+    DataTypeFunctions, Error, Module as _, RClass, TypedData, Value,
+};
+use rb_sys::{rb_ivar_set, rb_obj_freeze, rb_str_new_static};
+#[cfg(ruby_gte_3_0)]
+use rb_sys::{
+    rb_memory_view_entry_t, rb_memory_view_init_as_byte_array, rb_memory_view_register,
+    rb_memory_view_t, VALUE,
+};
+use std::ops::Range;
+
+/// @yard
+/// @rename Wasmtime::Memory::Slice
+/// Represents a slice of a WebAssembly memory. This is useful for creating Ruby
+/// strings from WASM memory without any extra memory allocations.
+#[derive(Debug)]
+pub struct Slice<'a> {
+    memory: MemoryGuard<'a>,
+    range: Range<usize>,
+}
+
+define_rb_intern!(IVAR_NAME => "@__slice__",);
+
+unsafe impl TypedData for Slice<'_> {
+    fn class() -> magnus::RClass {
+        *memoize!(RClass: define_data_class!(Memory::class(), "Slice"))
+    }
+
+    fn data_type() -> &'static magnus::DataType {
+        memoize!(magnus::DataType: {
+            let mut builder = DataTypeBuilder::<Slice>::new("Wasmtime::Memory::Slice");
+            builder.free_immediately();
+            builder.mark();
+            builder.build()
+        })
+    }
+}
+
+impl DataTypeFunctions for Slice<'_> {
+    fn mark(&self) {
+        self.memory.mark()
+    }
+}
+
+#[cfg(ruby_gte_3_0)]
+fn fiddle_memory_view_class() -> Option<RClass> {
+    let fiddle = object().const_get::<_, RModule>("Fiddle").ok()?;
+    fiddle.const_get("MemoryView").ok()
+}
+
+impl<'a> Slice<'a> {
+    pub fn new(memory: WrappedStruct<Memory<'a>>, range: Range<usize>) -> Result<Self, Error> {
+        let memory = MemoryGuard::new(memory)?;
+        let slice = Self { memory, range };
+        let _ = slice.get_raw_slice()?; // Check that the slice is valid.
+        Ok(slice)
+    }
+
+    /// @yard
+    /// Get this slice as a Fiddle memory view, which can be cheaply read by
+    /// other Ruby extensions.
+    ///
+    /// @def to_memory_view
+    /// @return [Fiddle::MemoryView] Memory view of the slice.
+    #[cfg(ruby_gte_3_0)]
+    pub fn to_memory_view(rb_self: WrappedStruct<Self>) -> Result<Value, Error> {
+        let klass = *memoize!(RClass: {
+            let c = fiddle_memory_view_class().unwrap();
+            gc::register_mark_object(c);
+            c
+        });
+
+        klass.new_instance((rb_self,))
+    }
+
+    /// @yard
+    /// Gets the memory slice as a Ruby string without copying the underlying buffer.
+    ///
+    /// @def to_str
+    /// @return [String] Binary +String+ of the memory.
+    pub fn to_str(rb_self: WrappedStruct<Self>) -> Result<Value, Error> {
+        let raw_slice = rb_self.get()?.get_raw_slice()?;
+        let id: Id = (*IVAR_NAME).into();
+        let rstring = unsafe {
+            let val = rb_str_new_static(raw_slice.as_ptr() as _, raw_slice.len() as _);
+            rb_ivar_set(val, id.as_raw(), rb_self.as_raw());
```

Comment:
Re-implementation of [`Fiddle::MemoryView#to_s`](https://github.com/ruby/fiddle/blob/36b24325754880266bacf49690e09826b30ad30b/ext/fiddle/memory_view.c#L271)